Dockerfile

```diff
@@ -1,93 +1,131 @@
+    # dev dependencies
+    binutils libproj-dev \
+    # debug tools
     curl net-tools iputils-ping postgresql-client-common jq openssh-client \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/* \
     && pip3 install pipenv \
     && pipenv --version
 
-COPY Pipfile* /tmp/
-RUN cd /tmp && \
-    pipenv install --system --deploy --ignore-pipfile
+ENV PIPENV_VENV_IN_PROJECT=1
+COPY Pipfile.lock Pipfile ./
+RUN pipenv sync --dev
+
+# Is it safe for the executable files to be writable by the user who executes them? (At a first
+# glance, it seems to also work without this chown, but as I am not sure why this was added I left
+# it like this for the moment)
+COPY --chown=${USER}:${GROUP} .env.default ./
+COPY --chown=${USER}:${GROUP} spec/ spec/
+COPY --chown=${USER}:${GROUP} app/ app/
+
+# Activate virtualenv
+ENV VIRTUAL_ENV=${INSTALL_DIR}/.venv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+
+# this is only used with the docker-compose setup within CI
+# to ensure that the app is only started once the DB container
+# is ready
+COPY ./wait-for-it.sh /app/
 
-# Set the working dir and copy the app
-WORKDIR /app
-COPY --chown=geoadmin:geoadmin .env.default /
-COPY --chown=geoadmin:geoadmin ./spec /spec/
-COPY --chown=geoadmin:geoadmin ./app /app/
+# Overwrite the version.py from source with the actual version
+ARG VERSION=unknown
+RUN echo "APP_VERSION = '$VERSION'" > app/config/version.py
 
 ARG GIT_HASH=unknown
 ARG GIT_BRANCH=unknown
 ARG GIT_DIRTY=""
 ARG AUTHOR=unknown
-ARG VERSION=unknown
-ARG TARGET=
 LABEL git.hash=$GIT_HASH
 LABEL git.branch=$GIT_BRANCH
 LABEL git.dirty="$GIT_DIRTY"
 LABEL author=$AUTHOR
 LABEL version=$VERSION
 
-# Overwrite the version.py from source with the actual version
-RUN echo "APP_VERSION = '$VERSION'" > /app/config/version.py
+WORKDIR ${INSTALL_DIR}/app/
+USER ${USER}
 
 ###########################################################
-# Container to perform tests/management/dev tasks
-FROM base as debug
-
-LABEL target=debug
-
-RUN cd /tmp && \
-    pipenv install --system --deploy --ignore-pipfile --dev
-
-# this is only used with the docker-compose setup within CI
-# to ensure that the app is only started once the DB container
-# is ready
-COPY ./wait-for-it.sh /app/
-
-# for testing/management, settings.py just imports settings_dev
-RUN echo "from .settings_dev import *" > /app/config/settings.py \
-    && chown geoadmin:geoadmin /app/config/settings.py
-
-# Collect static files, uses the .env.default settings to avoid django raising settings error
-RUN APP_ENV=default LOGGING_CFG=0 ./manage.py collectstatic --noinput
+# Builder container
+FROM base as builder
+RUN apt-get update -qq \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y gdal-bin \
+    # dev dependencies
+    binutils libproj-dev \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/* \
+    && pip3 install pipenv \
+    && pipenv --version
 
-USER geoadmin
+ENV PIPENV_VENV_IN_PROJECT=1
+COPY Pipfile.lock Pipfile ./
```

Comment:
Pipfiles should not be copied in the installation dir as they should not be part of the final production image.

Reply:
I decided to delete the Pipfile after installation as the alternative (put the Pipfile in another dir and manually move the venv after installation) seemed more error prone to me.

Reply:
Or more simply put the pipfiles in a source directory, install dependencies there and copy only the venv over the prod images:

```
# Builder image
# Virtual Env
ENV SRC_DIR=/usr/src
COPY Pipfile.lock Pipfile ${SRC_DIR}/
WORKDIR ${SRC_DIR}
RUN /root/.local/bin/pipenv sync 

# App
COPY --chown=${USER}:${GROUP} .env.default ${INSTALL_DIR}/
COPY --chown=${USER}:${GROUP} spec/ ${INSTALL_DIR}/spec/
COPY --chown=${USER}:${GROUP} app/ ${INSTALL_DIR}/app/

# PROD image
COPY --from=builder --chown=${USER}:${GROUP} ${SRC_DIR}/.venv ${INSTALL_DIR}/* ${INSTALL_DIR}/
```

Something like this (the wildcard in copy need to be checked if it works). 

But I would rather be more explicit on the installation location and what to install and avoid removing files at the end.

Reply:
I mean the builder image is not meant to run the application which means that the venv might be in another directory as the application.

Reply:
Ok I did it like this